### PR TITLE
[DRAFT] ORC-1463: [Java] Support Brotli codec

### LINKED ISSUE #1563: ORC-1463: Support brotli compression codec

### PR DESCRIPTION
**Pull request by deshanxiao**
### What changes were proposed in this pull request?
This PR is aimed to add Brotli codec defined by [RFC7932](https://www.ietf.org/rfc/rfc7932.txt).

### Why are the changes needed?
To support more codec in ORC. 

### How was this patch tested?
UT and CI pipeline.

This closes #1563

### CODE REVIEW

**Comment by deshanxiao** [comment]
`com.github.rdblue:brotli-codec` has no binary for aarch64. We can skip it.

File: java/pom.xml [hunk 1]
```diff
@@ -271,6 +272,12 @@
         <version>1.1.1</version>
         <optional>true</optional>
       </dependency>
+      <dependency>
+        <groupId>com.github.rdblue</groupId>
+        <artifactId>brotli-codec</artifactId>
+        <version>${brotli-codec.version}</version>
+        <optional>true</optional>
```
Comment (dongjoon-hyun) [inline java/pom.xml:279]: Oh, got it. Apache ORC community has only an `Optional` dependency.

File: java/shims/pom.xml [hunk 2]
```diff
@@ -42,6 +42,10 @@
       <groupId>org.slf4j</groupId>
       <artifactId>slf4j-api</artifactId>
     </dependency>
+    <dependency>
+      <groupId>com.github.rdblue</groupId>
+      <artifactId>brotli-codec</artifactId>
+    </dependency>
```
Comment (dongjoon-hyun) [inline java/shims/pom.xml:48]: Sorry, but this looks wrong to me. Please do not add the compression dependency to `shims` module.

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you for making a PR. However, I have a concern on the dependency management. Like Apache ORC 1.9, `shim` module should have only `Hadoop`-shim related dependency and eventually we want to slim or remove `shim` module by using a single Hadoop client library.

- https://mvnrepository.com/artifact/org.apache.orc/orc-shims/1.9.0

File: java/core/pom.xml [hunk 3]
```diff
@@ -72,6 +72,10 @@
       <groupId>org.threeten</groupId>
       <artifactId>threeten-extra</artifactId>
     </dependency>
+    <dependency>
+      <groupId>com.github.rdblue</groupId>
```
[the remark this answers is not in the document]
Comment (deshanxiao) [reply java/core/pom.xml:76]: I tried looking for more common repositories, but couldn't find them. Here is parquet pom file: https://github.com/apache/parquet-mr/blob/d2c3c6d2e761a17b75d56fd356a37a2f754072f7/parquet-hadoop/pom.xml#L207

I'm glad to update it if we can find a better one.

File: java/shims/pom.xml (hunk 2 above)
Comment (deshanxiao) [reply java/shims/pom.xml:48]: Got it, let me try to re-implement this part.

**Comment by deshanxiao** [comment]
It seems that pom file has some foramt errors. Using `mvn spotless:apply` can auto fix it.

**Comment by deshanxiao** [comment]
Disable unit test when arch is `aarch64`. 

Hi @dongjoon-hyun  Could you help me trigger the Workflow? Thanks~

**Comment by dongjoon-hyun** [comment]
I checked but it seems the workflow is triggered already, @deshanxiao .

**Comment by deshanxiao** [comment]
Thanks @guiyanakuang for trigger the workflow.

**Comment by deshanxiao** [comment]
Hi @dongjoon-hyun, the pipeline has passed. If you have any comments for the PR. Please ping me, thanks~

**Review by dongjoon-hyun — DISMISSED** [review]
@deshanxiao We have two issues still.
1. ARM64 Issue: https://issues.apache.org/jira/browse/PARQUET-1975
2. Apache Parquet has it as `runtime` dependency. I guess we need to do in the same way.

**Comment by dongjoon-hyun** [comment]
BTW, do you know why [HADOOP-13126](https://issues.apache.org/jira/browse/HADOOP-13126) is still open?

And, Hadoop JIRA seems to say the following. Could you double-check if there is no license issue or not?
> [Brotli is licensed with the MIT license](https://github.com/google/brotli/blob/master/LICENSE), and the [JNI library jbrotli is ALv2](https://github.com/MeteoGroup/jbrotli/blob/master/LICENSE).

**Comment by deshanxiao** [comment]
> BTW, do you know why [HADOOP-13126](https://issues.apache.org/jira/browse/HADOOP-13126) is still open?
> 

It seems that the reviewer think this modification is too large to cause the Hadoop PR freeze.
See https://github.com/apache/hadoop/pull/2723.

> And, Hadoop JIRA seems to say the following. Could you double-check if there is no license issue or not?
> 
> > [Brotli is licensed with the MIT license](https://github.com/google/brotli/blob/master/LICENSE), and the [JNI library jbrotli is ALv2](https://github.com/MeteoGroup/jbrotli/blob/master/LICENSE).

From the dependencies:
https://github.com/rdblue/brotli-codec/blob/9bbc2bcec1e7307e84552f28e90f20cb3099f189/build.gradle#L58

We are only indirectly referencing `jbrotli`. So the license requirement you mentioned is right.

BTW,  `brotli-codec` is just an adapter to meet with hadoop codec framework. We can use jbrotli directly. What do you think? @dongjoon-hyun

File: java/shims/src/java/org/apache/orc/impl/HadoopShims.java [hunk 4]
```diff
@@ -35,6 +35,7 @@ enum DirectCompressionType {
     ZLIB_NOHEADER,
     ZLIB,
     SNAPPY,
+    BROTLI,
```
Comment (wgtmac) [inline java/shims/src/java/org/apache/orc/impl/HadoopShims.java:38]: IIUC, shims here are used to workaround codec provided by Hadoop libraries of different versions. It seems that we do not need this line.

File: proto/orc_proto.proto [hunk 5]
```diff
@@ -386,6 +386,7 @@ enum CompressionKind {
   LZO = 3;
   LZ4 = 4;
   ZSTD = 5;
+  BROTLI = 6;
```
Comment (wgtmac) [inline proto/orc_proto.proto:389]: In order to check this in, we'd better update the website and the C++ code (at least throw an unimplemented exception on the reader side)

**Review by wgtmac — COMMENTED** [review]
Before diving into the detail implementation, I just want to make sure if brotli is widely used compared to others. Google itself does not provide an encoder implementation in pure Java. If the benefit it brings does not worth the effort of supporting it, I would rather not do this.

File: proto/orc_proto.proto (hunk 5 above)
Comment (deshanxiao) [reply proto/orc_proto.proto:389]: Thanks @wgtmac I will do it.

**Comment by deshanxiao** [comment]
> I found this one but not sure if it is mature enough: https://github.com/hyperxpro/Brotli4j

From 
https://issues.apache.org/jira/browse/PARQUET-1975?page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel&focusedCommentId=17281798#comment-17281798

Looks like someone else has had a similar idea. It seems that someone has mentioned a similar idea, but for some reason, parquet still uses a temporary maven library.

I checked that netty is using this library in:
https://github.com/netty/netty/blob/1c5053740dccc4ec952b4cfb8c4bcf697b49d9b6/pom.xml#L763

We could try to re-implement with it @wgtmac

File: java/shims/src/java/org/apache/orc/impl/HadoopShims.java (hunk 4 above)
Comment (deshanxiao) [reply java/shims/src/java/org/apache/orc/impl/HadoopShims.java:38]: Thanks I will remove it.

**Comment by deshanxiao** [comment]
> Before diving into the detail implementation, I just want to make sure if brotli is widely used compared to others. Google itself does not provide an encoder implementation in pure Java. If the benefit it brings does not worth the effort of supporting it, I would rather not do this.

From the test results of this [Jira](https://issues.apache.org/jira/browse/HADOOP-13126), I am positive about supporting brotli.

I add an [Issue](https://github.com/google/brotli/issues/1060) to track it.

**Comment by deshanxiao** [comment]
As mentioned in this https://github.com/google/brotli/issues/1060, we could wait for official support to come out.

I will pause this PR until its dependencies are ready, what do you think? @dongjoon-hyun @wgtmac

**Comment by dongjoon-hyun** [comment]
+1 for pausing. Thank you, @deshanxiao .

**Review by dongjoon-hyun — COMMENTED** [review]
To @deshanxiao and @wgtmac , here are my thoughts.

1. For `Parquet` feature parity, we need this.
2. Since this PR needs to change `orc_proto.proto`, we need to proceed before #1543 (orc-format v1.0.0) if we want to have this.
3. If there is no upstream change, the only way could be the `Parquet` way because this is for `Parquet`-feature parity.

- Parquet doesn't support this in arm64 architecture. Also, it's `optional` and `runtime` dependency. We should follow it if we want to proceed this.

```
  <profiles>
    <!-- PARQUET-1975 Do not add brotli-codec for ARM64 architectures -->
    <profile>
      <id>non-aarch64</id>
      <activation>
        <os>
          <arch>!aarch64</arch>
        </os>
      </activation>
      <dependencies>
        <dependency>
          <groupId>com.github.rdblue</groupId>
          <artifactId>brotli-codec</artifactId>
          <version>${brotli-codec.version}</version>
          <scope>runtime</scope>
          <optional>true</optional>
        </dependency>
      </dependencies>
    </profile>
  </profiles>
```

WDYT?

**Comment by deshanxiao** [comment]
+1 for add this to orc-format repo.

BTW, before implementing the feature, please give me some time to summarize the pros and cons of all third-party libraries.

**Comment by wgtmac** [comment]
Do we need a PoC implementation on the C++ side before making a decision? AFAIK, official C++ brotli library is complete and it is used in the Apache Arrow for years.

**Comment by dongjoon-hyun** [comment]
@wgtmac No~ This PR (Java) implementation is enough to add BROTLI to `orc-format`. C++ implementation can follow later.
> Do we need a PoC implementation on the C++ side before making a decision?

**Review by dongjoon-hyun — COMMENTED** [review]
FYI, `orc_proto.proto` will be deleted via the following.
- https://github.com/apache/orc/pull/1684

**Comment by wgtmac** [comment]
BTW, as parquet feature parity has been mentioned here, I have a couple of questions:
- Do we need to formally start the action?
- Should we also keep an eye on the lance format?
- Is it a good time to revisit the TODO items on ORC v2 (https://orc.apache.org/specification/ORCv2/)?

@dongjoon-hyun @mystic-lama

**Comment by dongjoon-hyun** [comment]
Could you send an email to dev@orc mailing list, @wgtmac ?

**Comment by dongjoon-hyun** [comment]
Okay, let me make a PR for you, @deshanxiao .

**Comment by dongjoon-hyun** [comment]
This is the PR.
- https://github.com/apache/orc-format/pull/7

**Comment by deshanxiao** [comment]
Thank you @dongjoon-hyun !

**Comment by dongjoon-hyun** [comment]
Do you think you can resume this, @deshanxiao ?

**Comment by deshanxiao** [comment]
> May I ask if you think you can resume this, @deshanxiao ?

Sure, this work will resume soon.

**Comment by deshanxiao** [comment]
The following is a comparison of the various implementation options we can take:

### Option 1: (Parquet mode):
Use the same way with parquet to support brotli codec, which will use a private maven repository with optional and runtime dependency.

#### Pros:
1.This method has been tested in practice and has been adopted by parquet.

#### Cons:
1. aarch64 is not supported.
2. We will have to use private maven repository. The code in this repository has not been updated in many years.

### Option 2: brotli4j (Netty mode):
#### Pros:
1. brotli4j is still under maintenance and supports all platforms.
2. license check is passed.

image


#### Cons:
1.This solution was mentioned in the parquet community, but in the end, Option 1 was adopted. The specific reason is unknown.



### Option 3: google/brotli:
#### Pros:
1.Official support.

#### Cons:
1.No brotli encoder. We need to wait for official support.


At present, I prefer `Option 2`. What do you think? @dongjoon-hyun @mystic-lama @wgtmac

**Comment by wgtmac** [comment]
Option 2 looks best on my side. Thanks @deshanxiao!

**Comment by dongjoon-hyun** [comment]
Thank you, @deshanxiao .  I agree with you. +1 for Option 2.

**Comment by deshanxiao** [comment]
Thank you @dongjoon-hyun @wgtmac . Please let me re-implement a version based on brotli4j.

Close this PR and will create a new one.

**Comment by dongjoon-hyun** [comment]
Thank you for leading this area, @deshanxiao .

**Comment by dongjoon-hyun** [comment]
cc @williamhyun too for the above context.

**Pull request closed without merging**